cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformExecutor.java

```diff
@@ -36,35 +36,45 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(TransformExecutor.class);
 
-  private final List<Transformation> transforms;
+  private final List<TransformationDetails> transforms;
   private final List<DefaultEmitter> emitters;
 
-  public TransformExecutor(List<Transformation> transforms, List<StageMetrics> transformMetrics) {
-    int numTransforms = transforms.size();
-    Preconditions.checkArgument(numTransforms == transformMetrics.size());
-    this.transforms = Lists.newArrayListWithCapacity(numTransforms);
+
+  public TransformExecutor(List<TransformationDetails> transformationDetailsList) {
+    int numTransforms = transformationDetailsList.size();
+    this.transforms = new ArrayList<>(numTransforms);
     this.emitters = Lists.newArrayListWithCapacity(numTransforms);
-    for (int i = 0; i < numTransforms; i++) {
-      StageMetrics stageMetrics = transformMetrics.get(i);
-      this.transforms.add(new TrackedTransform(transforms.get(i), stageMetrics));
-      this.emitters.add(new DefaultEmitter(stageMetrics));
+
+    for (TransformationDetails transformationDetails : transformationDetailsList) {
+      this.transforms.add(new TransformationDetails(transformationDetails,
+                                                    new TrackedTransform(transformationDetails.getTransformation(),
+                                                                         transformationDetails.getMetrics())));
+      this.emitters.add(new DefaultEmitter(transformationDetails.getMetrics()));
     }
   }
 
-  public Iterable<OUT> runOneIteration(IN input) throws Exception {
+  public TransformResponse<OUT> runOneIteration(IN input) throws Exception {
+    List<TransformResponse.TransformError<OUT>> errorIterators = new ArrayList<>();
     if (transforms.isEmpty()) {
-      return Lists.newArrayList((OUT) input);
+      return new TransformResponse<OUT>(Lists.newArrayList((OUT) input).iterator(), errorIterators);
     }
 
-    Transformation transform = transforms.get(0);
+    TransformationDetails transformationDetails = transforms.get(0);
     DefaultEmitter currentEmitter = emitters.get(0);
     currentEmitter.reset();
+    Transformation transform = transformationDetails.getTransformation();
     transform.transform(input, currentEmitter);
 
-    DefaultEmitter previousEmitter = currentEmitter;
+    // if there are errors in the iterator, we add it to the errorIterators list
+    if (currentEmitter.getErrorIterator().hasNext()) {
```

Comment:
ok ignore that previous comment since we want to write everything at the same time in one transaction.  We can still make this better and avoid all the copying that is done by passing in a List<TransformError> to DefaultEmitter and just have the emitter add to the list.